packages/cli/src/lib/bundler/server.ts

```diff
@@ -139,58 +161,99 @@ export async function serveBundle(options: ServeOptions) {
+        viteNodePolyfills(),
+        createHtmlPlugin({
+          entry: paths.targetEntry,
+          template: 'public/index.html',
+          inject: {
+            data: {
+              config: frontendConfig,
+              publicPath: config.output?.publicPath,
+            },
+          },
+        }),
+      ],
+      server: {
+        host,
+        port,
       },
-      https:
-        url.protocol === 'https:'
+      publicDir: paths.targetPublic,
+      root: paths.targetPath,
+    });
+  } else {
+    const compiler = webpack(config);
+
+    server = new WebpackDevServer(
+      {
+        hot: !process.env.CI,
+        devMiddleware: {
+          publicPath: config.output?.publicPath as string,
+          stats: 'errors-warnings',
+        },
+        static: paths.targetPublic
           ? {
-              cert: fullConfig.getString('app.https.certificate.cert'),
-              key: fullConfig.getString('app.https.certificate.key'),
+              publicPath: config.output?.publicPath as string,
+              directory: paths.targetPublic,
             }
-          : false,
-      host,
-      port,
-      proxy: targetPkg.proxy,
-      // When the dev server is behind a proxy, the host and public hostname differ
-      allowedHosts: [url.hostname],
-      client: {
-        webSocketURL: 'auto://0.0.0.0:0/ws',
-      },
-    } as any,
-    compiler as any,
-  );
-
-  await new Promise<void>((resolve, reject) => {
-    server?.startCallback((err?: Error) => {
-      if (err) {
-        reject(err);
-        return;
-      }
+          : undefined,
+        historyApiFallback: {
+          // Paths with dots should still use the history fallback.
+          // See https://github.com/facebookincubator/create-react-app/issues/387.
+          disableDotRule: true,
+
+          // The index needs to be rewritten relative to the new public path, including subroutes.
+          index: `${config.output?.publicPath}index.html`,
+        },
+        https:
+          url.protocol === 'https:'
+            ? {
+                cert: fullConfig.getString('app.https.certificate.cert'),
+                key: fullConfig.getString('app.https.certificate.key'),
+              }
+            : false,
+        host,
+        port,
+        proxy: targetPkg.proxy,
+        // When the dev server is behind a proxy, the host and public hostname differ
+        allowedHosts: [url.hostname],
+        client: {
+          webSocketURL: 'auto://0.0.0.0:0/ws',
+        },
+      } as any,
+      compiler as any,
+    );
+  }
 
-      openBrowser(url.href);
+  await new Promise<void>(async (resolve, reject) => {
+    if (process.env.EXPERIMENTAL_VITE) {
+      await (server as vite.ViteDevServer).listen();
       resolve();
-    });
+    } else {
+      (server as WebpackDevServer).startCallback((err?: Error) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+      });
+    }
+    openBrowser(url.href);
+    resolve();
```

Comment:
just noticed that the `resolve` function might be called twice at this point

Reply:
nice catch

Reply:
More importantly this changes the logic for the webpack server, I think it's best to duplicate this in the two if blocks tbh